crates/biome_service/src/projects.rs

```diff
@@ -133,37 +133,64 @@ impl Projects {
     }
 
     pub fn is_ignored_by_top_level_config(&self, project_key: ProjectKey, path: &Utf8Path) -> bool {
-        let is_included = self
-            .get_settings_based_on_path(project_key, path)
-            .is_some_and(|settings| {
-                let includes = &settings.files.includes;
-
-                let mut is_included = true;
-                if !includes.is_unset() {
-                    is_included = if is_dir(path) {
-                        includes.matches_directory_with_exceptions(path)
-                    } else {
-                        includes.matches_with_exceptions(path)
-                    };
-                }
+        self.is_ignored(project_key, path, FeatureName::empty())
+    }
 
-                is_included
-            });
-
-        // We store ignore matches inside the root settings, regardless of what package we are analyzing
-        let is_vcs_ignored = self.get_root_settings(project_key).is_some_and(|settings| {
-            if settings.vcs_settings.should_use_ignore_file() {
-                settings
-                    .vcs_settings
-                    .ignore_matches
-                    .as_ref()
-                    .is_some_and(|ignored_matches| ignored_matches.is_ignored(path, is_dir(path)))
+    pub fn is_ignored(
+        &self,
+        project_key: ProjectKey,
+        path: &Utf8Path,
+        features: FeatureName,
+    ) -> bool {
+        let data = self.0.pin();
+        let Some(project_data) = data.get(&project_key) else {
+            return false;
+        };
+
+        // First check if the path is ignored by the `files.includes` setting
+        // relevant to the given `path`.
+        let files_includes = project_data
+            .nested_settings
+            .iter()
+            .find(|(project_path, _)| path.starts_with(project_path))
+            .map_or(
+                &project_data.root_settings.files.includes,
+                |(_, settings)| &settings.files.includes,
+            );
+
+        let mut is_included = true;
+        if !files_includes.is_unset() {
+            is_included = if is_dir(path) {
+                files_includes.matches_directory_with_exceptions(path)
             } else {
-                false
-            }
-        });
+                files_includes.matches_with_exceptions(path)
+            };
+        }
 
-        !is_included || is_vcs_ignored
+        if !is_included {
+            return true;
+        }
```

Comment:
Careful here! I already did try this approach, and - you wouldn't believe it - but it regresses the performance a lot! I believe LLVM does some optimisations when both Booleans are returned at the end.

Just in case, compile for production, and test it again a big repository

Reply:
Just tried it, but it doesn't make a measurable difference for me either way. Of course the code is not the same anymore, and there is a third boolean now. Oh well, I looks a bit nicer without the early returns anyway :)